packages/shared/src/label-expression-selector/labelExpressionSelector.tsx

```diff
@@ -0,0 +1,424 @@
+  labels,
+  isValidationEnabled,
+  onSelect,
+}) => {
+  const { key, operator, values } = selectedExpression;
+
+  const { t } = useCustomTranslation();
+
+  // Display each label key as options
+  const keyOptions = React.useMemo(
+    () =>
+      Object.keys(labels).map((labelKey, i) => (
+        <SelectOption key={i} value={labelKey} />
+      )),
+    [labels]
+  );
+
+  // Default options of Operator enum.
+  const operatorOptions = Object.values(Operator).map((option, i) => (
+    <SelectOption key={i} value={option} />
+  ));
+
+  // Display each values of the selected key as options
+  // Modify value options based on key selection.
+  const valueOptions = React.useMemo(
+    () =>
+      (!!key ? labels[key] || [] : []).map((value, i) => (
+        <SelectOption key={i} value={value} />
+      )),
+    [labels, key]
+  );
+
+  // Reset values when label is selected.
+  const onKeyChange = React.useCallback(
+    (selectedKey: string) => {
+      onSelect(index, {
+        key: selectedKey,
+        operator,
+        values: key === selectedKey ? values : [],
+      });
+    },
+    [index, key, operator, values, onSelect]
+  );
+
+  // Reset values when label only operator is selected.
+  const onOperatorChange = React.useCallback(
+    (selectedOperator: Operator) => {
+      onSelect(index, {
+        ...selectedExpression,
+        operator: selectedOperator,
+        values: !isLabelOnlyOperator(selectedOperator) ? values : [],
+      });
+    },
+    [index, selectedExpression, values, onSelect]
+  );
+
+  // Call onSelect to update the state
+  const onValuesChange = React.useCallback(
+    (selectedValues: string[]) => {
+      onSelect(index, {
+        ...selectedExpression,
+        values: selectedValues,
+      });
+    },
+    [index, selectedExpression, onSelect]
+  );
+
+  const isKeyValid = getValidatedProp(isValidationEnabled && !key);
+  const isValuesValid = getValidatedProp(isValidationEnabled && !values.length);
+
+  return (
+    <Grid hasGutter>
+      <GridItem lg={4} sm={4}>
+        <FormGroup
+          label={t('Label')}
+          helperTextInvalid={t('Required')}
+          hasNoPaddingTop
+          isRequired
+          validated={isKeyValid}
+        >
+          <SingleSelectDropdown
+            data-test-id={`key-selection-dropdown-${index}`}
+            id="label-selection-dropdown"
+            placeholderText={t('Select a label')}
+            selectedKey={key}
+            selectOptions={keyOptions}
+            onChange={onKeyChange}
+            hasInlineFilter
+            required
+            validated={isKeyValid}
+            isCreatable
+          />
+        </FormGroup>
+      </GridItem>
+      <GridItem lg={4} sm={4}>
+        <FormGroup
+          label={t('Operator')}
+          helperTextInvalid={t('Required')}
+          isRequired
+          hasNoPaddingTop
+        >
+          <SingleSelectDropdown
+            data-test-id={`operator-selection-dropdown-${index}`}
```

Comment:
`data-test-id` is legacy convention...
```suggestion
            data-test={`operator-selection-dropdown-${index}`}
```